src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,144 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private T[] elements;
+    private int size = 0;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    private void ensureCapacity() {
+        T[] newElements = (T[]) new Object[elements.length * 2];
+        copyArray(elements, 0, newElements, 0, size);
+        elements = newElements;
+    }
+
+    private void copyArray(T[] arrayFrom,
+                           int arrayFromPos,
+                           T[] arrayTo,
+                           int arrayToPos,
+                           int amount) {
+        for (int i = 0; i < amount; i++) {
+            arrayTo[arrayToPos + i] = arrayFrom[arrayFromPos + i];
+        }
+    }
+
+    private boolean isFull() {
+        return size == elements.length;
+    }
+
+    private boolean isIndexValid(int index) {
+        return index < size && index >= 0;
+    }
+
+    private boolean isIndexValidForAdd(int index) {
+        return index <= size && index >= 0;
+    }
+
     @Override
     public void add(T value) {
-
+        elements[size] = value;
+        size++;
+        if (isFull()) {
+            ensureCapacity();
+        }
```

Comment:
Check if the array is full before adding the element. If you add the element first and then check if the array is full, you risk getting an ArrayIndexOutOfBoundsException.